src/servicesHost.ts

```diff
@@ -27,8 +27,17 @@ function makeServicesHost(
     const resolveSync = makeResolver(loader.options);
 
     const moduleResolutionHost = {
-        fileExists: (fileName: string) => utils.readFile(fileName) !== undefined,
-        readFile: (fileName: string) => utils.readFile(fileName),
+        fileExists: (fileName: string) => {
```

Comment:
I'll add some comments.

However I don't think it is necessary to wrap the code with a check for `usePreviousLoaderGeneratedFiles`. The `files` object is the one already used in ts-loader to cache files. Its contents are loaded in the following places:

* From the file system initially in `getTypeScriptInstance`:
```ts
const files: interfaces.TSFiles = {};
...
const filesToLoad = configParseResult.fileNames;
filesToLoad.forEach(filePath => {
	normalizedFilePath = path.normalize(filePath);
	files[normalizedFilePath] = {
		text: fs.readFileSync(normalizedFilePath, 'utf-8'),
		version: 0
	};
  });
```
* In the loader itself, using `updateFileInCache` (I've excluded my additional calls to `updateFileInCache` for this discussion):
```ts
const rawFilePath = path.normalize(this.resourcePath);
const filePath = utils.appendTsSuffixIfMatch(options.appendTsSuffixTo, rawFilePath);
const fileVersion = updateFileInCache(filePath, contents, instance);
```
* In `watch-run.ts` where file changes are registered

So it should be safe enough to load a file from `files` and use that, if it exists, instead of reading the file from the file system. This should even result in a slight performance improvement.

I would actually recommend updating `files` after reading them from the file system, and being able to have `text == undefined` for those files, that do not exist. This would reduce the file system access to once per file (initially - `watch-run` can of cause re-read any number of files).

All uses of `file.text` should be checked so they can handle `file.text == undefined`.

I am very happy to make this change - either in this PR or a new one, if you like.

There is one fly in the ointment. When I look at `watch-run.ts`, I see:
```ts
file.text = utils.readFile(filePath) || '';
```
This means that files that no longer exist are not deleted and have a valid `text`.